buildkitereporter/buildkitereporter.go

```diff
@@ -0,0 +1,168 @@
+		})
+		r.Subscribe(func(events <-chan taskrunner.ExecutorEvent) error {
+			for ev := range events {
+				switch event := ev.(type) {
+				case *taskrunner.TaskLogEvent:
+					if event.Stream == taskrunner.TaskLogEventStderr {
+						reporter.AppendLogs(event.TaskHandler().Definition(), event.Message)
+					}
+				}
+			}
+			return nil
+		})
+	}
+}
+
+type reporter struct {
+	sync.Mutex
+	path           string
+	uploadDirectly bool
+	maxLines       int
+	stderrs        map[*taskrunner.Task][]string
+}
+
+func newReporter() *reporter {
+	return &reporter{
+		maxLines: 50,
+		path:     "taskrunner.annotation.md",
+		stderrs:  make(map[*taskrunner.Task][]string),
+	}
+}
+
+func messageToLines(message string) []string {
+	return strings.Split(strings.TrimSpace(message), "\n")
+}
+
+func lastNElements(slice []string, n int) []string {
+	start := len(slice) - n
+	if start < 0 {
+		start = 0
+	}
+
+	return slice[start:]
+}
+
+// AppendLogs appends a log message to the reporter's cache, keeping at most N entries.
+func (r *reporter) AppendLogs(task *taskrunner.Task, message string) {
+	r.Lock()
+	defer r.Unlock()
+	r.stderrs[task] = lastNElements(
+		append(r.stderrs[task], messageToLines(message)...),
+		r.maxLines,
+	)
+}
+
+// Finish writes or uploads the buildkite annotation.
+func (r *reporter) Finish(tasks []*taskrunner.TaskHandler) error {
+	if r.uploadDirectly {
+		return r.upload(tasks)
+	}
+	return r.writeFile(tasks)
+}
+
+func (r *reporter) writeFile(tasks []*taskrunner.TaskHandler) error {
+	reader := r.write(tasks)
+
+	if err := os.MkdirAll(filepath.Dir(r.path), 0755); err != nil {
+		return err
+	}
+
+	f, err := os.OpenFile(r.path, os.O_WRONLY|os.O_CREATE, 0660)
+	defer f.Close()
+	if err != nil {
+		return err
+	}
+
+	if _, err := io.Copy(f, reader); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *reporter) upload(tasks []*taskrunner.TaskHandler) error {
+	reader := r.write(tasks)
+	cmd := exec.Command("buildkite-agent", "annotate", "--style=error", "--context=taskrunner", "--append")
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return err
+	}
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	if _, err := io.Copy(stdin, reader); err != nil {
+		return err
+	}
+	if err := stdin.Close(); err != nil {
+		return err
+	}
+
+	if err := cmd.Wait(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// writeFile creates a buildkite annotation.
+func (r *reporter) write(tasks []*taskrunner.TaskHandler) io.Reader {
+	var buf bytes.Buffer
+	var failureCount int
+	var failedTasks []*taskrunner.Task
+
+	for _, task := range tasks {
+		if task.State() == taskrunner.TaskHandlerExecutionState_Error {
+			failureCount++
+			failedTasks = append(failedTasks, task.Definition())
+		}
+	}
+
+	fmt.Fprintf(&buf, "<p>taskrunner: %d tasks failed</p>\n", failureCount)
+
+	for _, task := range failedTasks {
+		fmt.Fprintf(&buf, "<details><summary><code>%s</code></summary>\n<pre>", task.Name)
```

Comment:
is this a buildkite format? or junit? could you link to a reference in a comment?

Reply:
It's super poorly documented but it's an annotation format.

Reply:
could you link to the (even poor) documentation?

Reply:
Done - both in commit and file